rangeserver/src/cache/memtabledb.rs

```diff
@@ -0,0 +1,648 @@
+use skiplist::OrderedSkipList;
+use std::cmp::Ordering;
+use tokio::sync::RwLock;
+use std::mem;
+use bytes::Bytes;
+use std::sync::Arc;
+use crate::cache::{Cache, Error, GCCallback, CacheOptions};
+use std::collections::HashMap;
+use std::vec::Vec;
+use std::thread;
+use std::sync::Mutex;
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use std::cmp::min;
+use std::cmp::max;
+use core::ops::Bound::Included;
+
+// memtable status changes from FLUSHED->MUTABLE->IMMUTABLE->FLUSHED
+#[derive(Clone, Debug, PartialEq)]
+pub enum MemTableStatus {
+    FLUSHED,    // also initial status
+    MUTABLE,
+    IMMUTABLE,
+}
+
+#[derive(Clone, Debug, Eq)]
+pub struct MemTableEntry {
+    key: Bytes,
+    val: Bytes,
+    epoch: u64, 
+    deleted: bool,
+}
+
+impl MemTableEntry {
+    pub fn new(key: Bytes, val: Bytes, epoch: u64, deleted: bool, ) -> Self {
+        MemTableEntry { key, val, epoch, deleted }
+    }
+}
+
+impl PartialOrd for MemTableEntry {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for MemTableEntry {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let key_comp = self.key.cmp(&other.key);
+        match key_comp {
+            Ordering::Less | Ordering::Greater => key_comp,
+            Ordering::Equal => other.epoch.cmp(&self.epoch),
+        }
+    }
+}
+
+impl PartialEq for MemTableEntry {
+    fn eq(&self, other: &Self) -> bool {
+        self.key == other.key && self.epoch == other.epoch
+    }
+}
+
+#[derive(Debug)]
+pub struct MemTable {
+    id: usize,
+    size_in_bytes: u64,
+    status: MemTableStatus,
+    latest_epoch: u64,
+    oldest_epoch: u64,
+    entries: OrderedSkipList<MemTableEntry>,
```

Comment:
I was thinking about the data structure choice a bit more. It looks like we don't currently have any scan operation to return keys in order. Is that planned for the future? Otherwise a hash table of ordered lists would be better where each key maps to an ordered list of epoch values. Anyways, it would be good to document why this data structure is right for the use-case in a comment.

Reply:
I believe scan is a use case. The first cut implementation doesn't have it but we will need it. 
@tamer-eldeeb what do you think?

Reply:
yes we generally need ordered data structures to support range reads